pkg/broker/receiver.go

```diff
@@ -237,37 +237,84 @@ func (r *Receiver) getTrigger(ctx context.Context, ref path.NamespacedNameUID) (
+	// Record event count and filtering time.
 	startTS := time.Now()
 	defer func() {
 		filterTimeMS := int64(time.Now().Sub(startTS) / time.Millisecond)
 		stats.Record(ctx, MeasureTriggerFilterTime.M(filterTimeMS))
 	}()
 
-	filterType := ts.Filter.SourceAndType.Type
-	if filterType != eventingv1alpha1.TriggerAnyFilter && filterType != event.Type() {
-		r.logger.Debug("Wrong type", zap.String("trigger.spec.filter.sourceAndType.type", filterType), zap.String("event.Type()", event.Type()))
-		ctx, _ = tag.New(ctx, tag.Upsert(TagFilterResult, "fail"))
-		return false
+	// No filter specified, default to passing everything.
+	if ts.Filter.DeprecatedSourceAndType == nil && ts.Filter.Attributes == nil {
+		ctx, _ = tag.New(ctx, tag.Upsert(TagFilterResult, "empty-pass"))
+		return true
 	}
-	filterSource := ts.Filter.SourceAndType.Source
-	s := event.Context.AsV01().Source
-	actualSource := s.String()
-	if filterSource != eventingv1alpha1.TriggerAnyFilter && filterSource != actualSource {
-		r.logger.Debug("Wrong source", zap.String("trigger.spec.filter.sourceAndType.source", filterSource), zap.String("message.source", actualSource))
+
+	attrs := map[string]string{}
+	// Since the filters cannot distinguish presence, filtering for an empty
+	// string is impossible.
+	if ts.Filter.DeprecatedSourceAndType != nil {
+		attrs["type"] = ts.Filter.DeprecatedSourceAndType.Type
+		attrs["source"] = ts.Filter.DeprecatedSourceAndType.Source
+	} else if ts.Filter.Attributes != nil {
+		attrs = map[string]string(*ts.Filter.Attributes)
+	}
+
+	result := r.filterEventByAttributes(attrs, event)
+	if result {
+		ctx, _ = tag.New(ctx, tag.Upsert(TagFilterResult, "pass"))
+	} else {
 		ctx, _ = tag.New(ctx, tag.Upsert(TagFilterResult, "fail"))
+	}
+	return result
+}
 
-		return false
+// filterEventByAttributes
+func (r *Receiver) filterEventByAttributes(attrs map[string]string, event *cloudevents.Event) bool {
+	// Set standard context attributes. The attributes available may not be
+	// exactly the same as the attributes defined in the current version of the
+	// CloudEvents spec.
+	ce := map[string]interface{}{
+		"specversion":         event.SpecVersion(),
+		"type":                event.Type(),
+		"source":              event.Source(),
+		"subject":             event.Subject(),
+		"id":                  event.ID(),
+		"time":                event.Time().String(),
+		"schemaurl":           event.SchemaURL(),
+		"datacontenttype":     event.DataContentType(),
+		"datamediatype":       event.DataMediaType(),
+		"datacontentencoding": event.DataContentEncoding(),
+	}
+	ext := event.Extensions()
+	if ext != nil {
+		for k, v := range ext {
+			ce[k] = v
+		}
 	}
 
-	ctx, _ = tag.New(ctx, tag.Upsert(TagFilterResult, "pass"))
+	for k, v := range attrs {
+		var value interface{}
+		value, ok := ce[k]
+		// If the attribute does not exist in the event, return false.
+		if !ok {
+			r.logger.Debug("No attribute", zap.String("attribute", k))
```

Comment:
```suggestion
			logging.FromContext(ctx).Debug("Attribute not found", zap.String("attribute", k))
```

and add `ctx` to this method. 

My reasoning is that the context is likely to have additional useful information about the event we are processing, which will make debugging much easier once we have many events running through a single process in parallel.

Reply:
Done, thanks!